server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/http/SessionSupplier.scala

```diff
@@ -0,0 +1,108 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ * ***************************************************************/
+
+package org.apache.james.jmap.http
+
+import java.net.URL
+
+import com.google.common.annotations.VisibleForTesting
+import eu.timepit.refined.auto._
+import eu.timepit.refined.refineV
+import org.apache.james.core.Username
+import org.apache.james.jmap.http.SessionSupplier.{CORE_CAPABILITY, MAIL_CAPABILITY}
+import org.apache.james.jmap.model.CapabilityIdentifier.CapabilityIdentifier
+import org.apache.james.jmap.model.Id.Id
+import org.apache.james.jmap.model._
+import reactor.core.publisher.Mono
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+object SessionSupplier {
+  private val CORE_CAPABILITY = CoreCapability(
+    properties = CoreCapabilityProperties(
+      MaxSizeUpload(10_000_000L),
+      MaxConcurrentUpload(4L),
+      MaxSizeRequest(10_000_000L),
+      MaxConcurrentRequests(4L),
+      MaxCallsInRequest(16L),
+      MaxObjectsInGet(500L),
+      MaxObjectsInSet(500L),
+      collationAlgorithms = List("i;unicode-casemap")))
+
+  private val MAIL_CAPABILITY = MailCapability(
+    properties = MailCapabilityProperties(
+      MaxMailboxesPerEmail(Some(10_000_000L)),
+      MaxMailboxDepth(None),
+      MaxSizeMailboxName(200L),
+      MaxSizeAttachmentsPerEmail(20_000_000L),
+      emailQuerySortOptions = List("receivedAt", "cc", "from", "to", "subject", "size", "sentAt", "hasKeyword", "uid", "Id"),
+      MayCreateTopLevelMailbox(true)
+    ))
+}
+
+class SessionSupplier {
+  def generate(username: Username): SMono[Session] =
+    SMono.fromPublisher(
+      Mono.zip(
+        accounts(username).asJava(),
+        primaryAccounts(username).asJava()))
+      .map(tuple => generate(username, tuple.getT1, tuple.getT2))
+
+  private def accounts(username: Username): SMono[Map[Id, Account]] =
+    getId(username)
+      .map(id => Map(
+        id -> Account(
+          username,
+          IsPersonal(true),
+          IsReadOnly(false),
+          accountCapabilities = Set(CORE_CAPABILITY, MAIL_CAPABILITY))))
+
+  private def primaryAccounts(username: Username): SMono[Map[CapabilityIdentifier, Id]] =
+    SFlux.just(CORE_CAPABILITY, MAIL_CAPABILITY)
+      .flatMap(capability => getId(username)
+        .map(id => (capability.identifier, id)))
+      .collectMap(getIdentifier, getId)
+  private def getIdentifier(tuple : (CapabilityIdentifier, Id)): CapabilityIdentifier = tuple._1
+  private def getId(tuple : (CapabilityIdentifier, Id)): Id = tuple._2
+
+  private def getId(username: Username): SMono[Id] = {
+    SMono.fromCallable(() => refineId(username))
+      .flatMap {
+        case Left(errorMessage: String) => SMono.raiseError(new IllegalStateException(errorMessage))
```

Comment:
It's not a state problem, it's an argument problem

Reply:
this code is removed